### PR TITLE
Include numeric subscripts and superscripts in symbols

### PR DESCRIPTION
**Pull request by zkirby**
Please see #41 for further discussion. Numbers are now considered symbols if and only if they are a direct descendent of a super script or sub script

### CODE REVIEW

**Comment by andrewhead** [comment]
Thanks for linking to the issue!

File: data-processing/common/parse_equation.py [hunk 1]
```diff
@@ -53,9 +57,12 @@ def get_symbols(mathml: str) -> List[Symbol]:
     node_symbols = []
     consecutive_char_sequence: List[BeautifulSoup] = [] # The symbol_nodes of consecutive characters
     consecutive_char_indices: List[int] = [] # Indices of the characters of a consecutive character sequence
-    # It's 'safe' to call soup.find_all as the underlying bs4 implementation 
+    # It's 'safe' to call soup.find_all as the underlying bs4 implementation
```
Comment (andrewhead) [inline data-processing/common/parse_equation.py]: Hahaha I'm just noticing the foreboding quotes around 'safe' here XD

File: data-processing/common/parse_equation.py [hunk 2]
```diff
@@ -53,9 +57,12 @@ def get_symbols(mathml: str) -> List[Symbol]:
...
     # appears to preserve order https://bazaar.launchpad.net/~leonardr/beautifulsoup/bs4/view/head:/bs4/element.py
     for symbol_node in soup.find_all(SYMBOL_TAGS):
+        # Skip partial symbols that aren't themselves descendants of another symbol
+        if symbol_node.name in PARTIAL_SYMBOLS and not symbol_node.parent.name in SYMBOL_TAGS:
+            continue
```
Comment (andrewhead) [inline data-processing/common/parse_equation.py]: I like the brevity, kudos! Isn't another way of implementing this just adding `mn` to the list of `CHARACTER_TAGS`, or am I missing something? One advantage of that approach is that characters are already another form of "partial symbols" (i.e. they're the pieces symbols are made up of), so then we only one data structure to hold 'em all.

Comment (zkirby) [reply data-processing/common/parse_equation.py]: Ah I see hahah, I think I misunderstood what characters were and what exactly symbols were but it appears that yes that is all that is needed LOL

File: data-processing/common/parse_equation.py (hunk 1 above)
Comment (zkirby) [reply data-processing/common/parse_equation.py]: XD

File: data-processing/common/parse_equation.py (hunk 2 above)
Comment (andrewhead) [reply data-processing/common/parse_equation.py]: Cool! My bad for my lackluster documentation.

File: data-processing/tests/mathml/x_sub_four_sub_three.xml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+            <mrow>
+                <msub s2:index="5">
+                    <mi s2:start="0" s2:end="1" s2:index="0">x</mi>
+                    <msub s2:index="4" s2:start="2" s2:end="7">
+                        <mn s2:start="3" s2:end="4" s2:index="1">4</mn>
+                        <mn s2:start="5" s2:end="6" s2:index="2">3</mn>
+                    </msub>
+                </msub>
+            </mrow>
+            <annotation encoding="application/x-tex">x_{t_i}</annotation>
```
Comment (andrewhead) [inline data-processing/tests/mathml/x_sub_four_sub_three.xml]: nit pick: can you update this element to have the value `x_43`? It's not important for the sake of the test running, though might avoid future confusion about this test data.

File: data-processing/tests/test_parse_equation.py [hunk 4]
```diff
@@ -86,6 +86,22 @@ def test_get_symbol_children():
...
+        symbols = get_symbols(mathml)
+
+    assert len(symbols) == 3
+    x_sub_four_sub_three = list(
+        filter(lambda s: "msub" in s.mathml and "x" in s.mathml, symbols)
+    )[0]
+    x = list(filter(lambda s: s.mathml == "<mi>x</mi>", symbols))[0]
+
+    assert len(x_sub_four_sub_three.children) == 2
+    assert x in x_sub_four_sub_three.children
```
Comment (andrewhead) [inline data-processing/tests/test_parse_equation.py]: Can you add an assertion to check here that the nodes `4` and `3` are among in `x_sub_four_sub_three.characters`?

File: data-processing/tests/test_parse_equation.py [hunk 5]
```diff
@@ -86,6 +86,22 @@ def test_get_symbol_children():
...
     assert i in t_sub_i.children
 
+def test_get_symbol_sub_super_with_numeric_child():
+    with open(
+        get_test_path(os.path.join("mathml", "x_sub_four_sub_three.xml"))
+    ) as mathml_file:
+        mathml = mathml_file.read()
+        symbols = get_symbols(mathml)
+
+    assert len(symbols) == 3
```
Comment (andrewhead) [inline data-processing/tests/test_parse_equation.py]: This is a great test. It makes me realize, do we want `4_3` to be detected as a symbol?

I feel like I we probably don't want `4^3` to be detected as a symbol---that's just a number. `4_3` has no special meaning, and probably won't come up in papers. Though I'm curious what your thoughts are. I'm thinking we may want to add a rule to skip `msub` and `msup` elements that have an `mn` as their base.

Comment (zkirby) [reply data-processing/tests/test_parse_equation.py]: I think I mostly agree. However, I am interested to know the situations in which 4_3 (or for that matter any calculation involving just numeric symbols) would not have been reduced? My gut says that if there is an expression that has intentionally not been reduced, it holds a special significance in the paper and thus might be something someone wants to actually look for. I'm unsure, what do you think?

Comment (andrewhead) [reply data-processing/tests/test_parse_equation.py]: That's fair! Okay, then let's keep numbers that have a superscript or subscript. We can decide when testing out the tool what feels right.

**Comment by zkirby** [comment]
Doing that right now, wanted to wait to see what we wanted to do for the fully numeric sub/sup :)

**Pull request merged**